stl/inc/format

```diff
@@ -389,26 +555,256 @@ public:
+
+    template <class _Id>
+    constexpr void _On_dynamic_width(_Id _Arg_id) {
+        this->_Specs._Width = _Get_dynamic_specs<_Width_checker>(_Get_arg(_Arg_id));
+    }
+
+    template <class _Id>
+    constexpr void _On_dynamic_precision(_Id _Arg_id) {
+        this->_Specs._Precision = _Get_dynamic_specs<_Precision_checker>(_Get_arg(_Arg_id));
+    }
+
+private:
+    _ParseContext& _Parse_ctx;
+    _Context& _Ctx;
+
+    constexpr basic_format_arg<_Context> _Get_arg(_Auto_id_tag) {
+        return _STD _Get_arg(_Ctx, _Parse_ctx.next_arg_id());
+    }
+
+    constexpr basic_format_arg<_Context> _Get_arg(size_t _Arg_id) {
+        _Parse_ctx.check_arg_id(_Arg_id);
+        return _STD _Get_arg(_Ctx, _Arg_id);
+    }
 };
+class _Numeric_specs_checker {
+    _Type _Arg_type = _Type::_None;
+
+public:
+    constexpr _Numeric_specs_checker(_Type _Arg_type) : _Arg_type(_Arg_type) {}
+
+    constexpr void _Require_numeric_argument() {
+        if (!_Is_numeric_fmt_type(_Arg_type)) {
+            throw format_error("Format specifier requires numeric argument.");
+        }
+    }
+
+    constexpr void _Check_sign() {
+        _Require_numeric_argument();
+        if (_Is_integral_fmt_type(_Arg_type) && _Arg_type != _Type::_Int32 && _Arg_type != _Type::_Int64
+            && _Arg_type != _Type::_Char) {
+            throw format_error("Format specifier requires signed argument.");
+        }
+    }
+
+    constexpr void _Check_precision() {
+        if (_Is_integral_fmt_type(_Arg_type) || _Arg_type == _Type::_Pointer) {
+            throw format_error("Precision not allowed for this argument type.");
+        }
+    }
+};
+
+// Uses _Numeric_specs_checker to check that the type of the argument printed by
+// a replacement field with format specs actually satisfies the requirements for
+// that format spec. If the requirements are met then calls the base class
+// handler method.
+template <class _Handler>
+class _Specs_checker : public _Handler {
+    _Numeric_specs_checker _Numeric_checker;
+
+public:
+    constexpr _Specs_checker(const _Handler& _Handler_inst, _Type _Arg_type)
+        : _Handler(_Handler_inst), _Numeric_checker(_Arg_type) {}
+
+    // _On_align has no checking, since we don't implement numeric alignments.
+    constexpr void _On_sign(_Sign _Sgn) {
+        _Numeric_checker._Check_sign();
+        _Handler::_On_sign(_Sgn);
+    }
+
+    constexpr void _On_hash() {
+        _Numeric_checker._Require_numeric_argument();
+        _Handler::_On_hash();
+    }
+
+    constexpr void _On_zero() {
+        _Numeric_checker._Require_numeric_argument();
+        _Handler::_On_zero();
+    }
+
+    constexpr void _On_precision(unsigned int _Precision) {
+        _Numeric_checker._Check_precision();
+        _Handler::_On_precision(_Precision);
+    }
+};
+
+template <class _Ty, class _CharT = char>
+struct formatter;
+
 
 // TODO: test coverage
 template <class _Context, class... _Args>
 class _Format_arg_store {

```

Comment:
The standard says

      template<class Context, class... Args> struct format-arg-store;      // exposition only

Does that permit an evil program to match its template parameter list with that of a template template argument?

Reply:
Yes, but I'm willing to fix the Standard if something in the wording prevents us making our implementation as efficient as otherwise possible for no good reason. (This comment is likely moot anyway given #1625.)